proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+   we gain more experience here: adding new features is easier than removing them.
+   While I agree that this could be done, the ``@(..)`` construct makes for a very
+   easy migration from today's ``-XScopedTypeVariables`` and is thus tempting to
+   be around from the start. I don't feel strongly.
+
+#. (technical) The `Visible Type Applications`_ (VTA) paper defines the behavior about what to
+   do when checking against a polytype: it says to deeply skolemize. However, eager deep
+   skolemization will spell trouble for this extension, as we need the lambdas to see
+   the ``forall``\s. The end of the Section 6.1 in the `extended VTA <https://cs.brynmawr.edu/~rae/papers/2016/type-app/visible-type-app-extended.pdf>`_ paper discusses
+   why we do eager deep skolemization: essentially, the alternative would be to do
+   type generalization at inflection points between checking and inference mode,
+   right before doing the subsumption check. Type generalization is hard in GHC, though,
+   and so the paper avoided it. In order to implement this proposal, we'll have to work
+   out how to do this.
+
+Costs and Drawbacks
+^^^^^^^^^^^^^^^^^^^
+
+1. This part of the proposal
+   is *not* backward-compatible with today's ``-XScopedTypeVariables``,
+   because it rejects expressions like ::
+
+     ((\x -> (x :: a)) :: forall a. a -> a)
+
+   which are accepted today. No migration period is proposed, because it is
+   very hard to imagine how ``-XTypeAbstractions`` and ``-XScopedForAlls`` should
+   co-exist peacefully here. Instead, we can issue a specific error message telling
+   users how to migrate their code in this case.
+
+   My hope is that constructs such as this one are rare and would not impact many
+   users.
+
+   If necessary, we could imagine taking the expression ``expr :: forall ... . ty``
+   and looking proactively to see whether ``expr`` ever uses a type variable
+   pattern from this proposal. If not, ``-XScopedForAlls`` could trigger (and we
+   issue a warning with ``-Wcompat``). But, if a type argument appears anywhere
+   in ``expr``, then ``-XScopedForAlls`` is disabled. This would be backward-compatible,
+   but unfortunately non-local and annoying. I prefer just to skip this
+   migration step.
+
+``let``-binding types
+~~~~~~~~~~~~~~~~~~~~~
+
+.. _type-let:
+
+This segment of the proposal goes beyond previous proposals in describing a mechanism
+to use ``let`` to bind type synonyms.
+
+Motivation
+^^^^^^^^^^
+
+1. Users have, from time to time, requested the ability to make local type synonyms.
+   GHC even has a little support for synonyms via equality constraints (e.g., writing
+   ``f :: (a ~ Some Big Type With Lots Of Parts) => Maybe a -> a -> Maybe a``). Instead
+   of encoding this idea via equality constraints, though, it would be nice to support
+   it directly.
+
+#. Now that type variables can stand for types, we can write code like ::
+
+     f :: Maybe Bool -> Bool
+     f (x :: Maybe b) = (True :: b)
```

Comment:
Why “Now that”? This is not introduced by this proposal, is it?

Reply:
No, it is not, but it still feels recent (to me). I've reworded to avoid this confusion.